website/docs/reference/cli/global-options.md
## `--api-listen-address` or `OPCTL_API_LISTEN_ADDRESS` *default: 127.0.42.224:80*
To specify the HOST:PORT on which the opctl API server will listen, include a `--api-listen-address` or set an `OPCTL_API_LISTEN_ADDRESS` env var.

### Examples
```sh
opctl --api-listen-address 127.0.0.1:8080
```

## `--container-runtime` or `OPCTL_CONTAINER_RUNTIME` *default: docker*
To specify the container runtime opctl uses to run containers, include a `--container-runtime` or set an `OPCTL_CONTAINER_RUNTIME` env var.

Allowed values are:
- `k8s` (connects to k8s via downward API)
- `docker` (connects to docker via same config as docker client)
- `qemu` (experimental)

### Examples
```sh
opctl --container-runtime qemu
```

## `--data-dir` or `OPCTL_DATA_DIR` *default: OS dependent per user app data*
To specify the path of the directory used to store opctl data, include a `--data-dir` or set an `OPCTL_DATA_DIR` env var.
to a relative or absolute path. 

### Examples
```sh
opctl --data-dir . node create
```

```sh
export OPCTL_DATA_DIR=. && opctl node create
```

## `--dns-listen-address` or `OPCTL_DNS_LISTEN_ADDRESS` *default: 127.0.0.1:53*
To specify the HOST:PORT on which the opctl DNS server will listen, include a `--dns-listen-address` or set an `OPCTL_DNS_LISTEN_ADDRESS` env var.

### Examples
```sh
opctl --dns-listen-address 0.0.0.0:53
```

## `-h` or `--help`
For context specific help, include a `-h` (or `--help`) flag w/ your command.

### Examples
```sh
opctl node create -h

Usage: opctl node create

Creates a node
```

## `--nc`
To disable color, include a `--nc` flag w/ your command.
> this may increase readability in environments not supporting color escape codes or piping output to another program.

### Examples
```sh
opctl --no-color events
```

## `-v` or `--version`
To print the version and exit, specify a `-v` (or `--version`) flag.

### Examples
```sh
opctl -v
```

docs/reference/op.yml/schema/run/container/envVars/envVar.md
An environment variable.

The value MUST be one of:

- `expression` coercible to string value
- `null` (will be bound to in scope ref w/ same name)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file in
accordance with
[![keepachangelog 1.0.0](https://img.shields.io/badge/keepachangelog-1.0.0-brightgreen.svg)](http://keepachangelog.com/en/1.0.0/)

## [0.1.64] - 2024-02-10

### Added

- Leverage native privilege escalation handling for self-update

## [0.1.63] - 2024-02-08

### Added

- Native privilege escalation handling (i.e. no more calling with sudo)

## [0.1.62] - 2024-02-05

### Fixed

- Put in a short term fix so that when `pullCreds` is used on Linux against
something other than `docker.io`, things don't error out any more.

## [0.1.60] - 2024-11-24

### Fixed

- Fix a regression in the 'opctl op install' command introduced in 0.1.48 which caused it to be much slower than previous
- Fix a potential race condition encountered when pulling/using multiple remote ops from the same repo at once

## [0.1.59] - 2024-11-17

### Fixed

- `No such container: opctl_[...]` errors if container exits too fast (race condition)

## [0.1.58] - 2024-11-13

### Fixed

- Fix executable bit not maintained within remote ops

## [0.1.57] - 2024-11-12

### Changed

- Don't log "not found in graph" to logs (noisy without value)

### Fixed

- Fix 'opctl node kill' panics if you've never run an op
- Rate limit errors surfacing as "image not found" errors
- Fixed file initialization results in incorrect (empty) file data

## [0.1.56] - 2024-11-11

### Added

- Access containers by their name from opctl nodes

### Deprecated

- `container.ports`; access containers by name instead

## [0.1.56-alpha.1] - 2024-11-08

### Added

- Access containers by their name from opctl nodes

### Deprecated

- `container.ports`; access containers by name instead

## [0.1.56-alpha.0] - 2024-10-30

### Added

- Access containers by their name from opctl nodes

### Deprecated

- `container.ports`; access containers by name instead

## [0.1.55] - 2024-09-04

### Fixed

- Fix long time to resolve non-existent op

## [0.1.53] - 2024-04-18

### Fixed

- A few dependency updates
  - in the Golang SDK the only breaking upgrade is Docker v23 -> v25.0.3+incompatible
  - in the JS SDK nock (a devDependency) was updated from 9 -> 13.
  - in the JS SDK react-ace was updated from 9 -> 11

### Removed

- Deleted the React SDK because it is unused

## [0.1.52] - 2023-06-06

### Fixed

- [Escaped references are not escaped if an unescaped '$' exists prior in the string](https://github.com/opctl/opctl/issues/1063)

## [0.1.51] - 2023-06-01

### Added

- [Variable reference as `container.cmd`](https://github.com/opctl/opctl/issues/1064)

### Fixed

- Simultaneously defining a default plus constraints on an object or array input results in a validation error

## [0.1.50] - 2023-05-01

### Added

- Automatic GPU detection/passthrough for docker runtime

### Fixed

- `opctl ls` renders ops with no description as blank line
- `lte` predicate
- In some cases, stacktraces logged as bytes rather than strings
- Containers aren't automatically removed after exit for docker runtime
- In some cases, `Runtime error: invalid memory address or nil pointer dereference` error occurs
- CLI not logging errors from the root op

## [0.1.49] - 2022-09-21

### Added

- opspec now supports gt, gte, lt, lte predicates
- `opctl node kill` will now stop and remove any opctl managed containers
- introduced `opctl node delete` command which "Deletes a node. This is destructive! all node data including auth, caches, and operation state will be permanently removed."

### Changed

- upgrading to this version from prior versions is destructive! all node data including auth, caches, and operation state will be permanently removed.
- K8s container runtime now explicitly deletes terminated pods

### Fixed

- [Node locking mechanism doesn't ensure process is opctl](https://github.com/opctl/opctl/issues/913)
- [CLI no longer logs errors occurring in parallel calls](https://github.com/opctl/opctl/issues/1032)

## [0.1.48] - 2021-08-13

### Added

- When running an op via opctl run, display progress via a live call graph
- When running an op via opctl run, prefix log lines emitted by workloads with their op id & ref
- Basic support for sending local files and directories to remote nodes when using the API client
- [Allow defining description on call graph nodes](https://github.com/opctl/opctl/issues/900)

### Changed

- Self-update now uses github releases instead of equinox.io
- API now limits request body to 40Mb
- [Improved error output when op resolution fails. You'll now see a list of resolutions tried and why each failed.](https://github.com/opctl/opctl/pull/883)
- [More consistent error messaging formats](https://github.com/opctl/opctl/pull/885)
- [Detect invalid op output names](https://github.com/opctl/opctl/issues/798)
- [Allow using type initializers in input/output defaults](https://github.com/opctl/opctl/issues/957)
- [Deprecated absolute paths as file/dir input/output defaults](https://github.com/opctl/opctl/issues/957)
- [Deprecated op output binding syntax; use same syntax as binding inputs](https://github.com/opctl/opctl/issues/721)
- [Deprecated param.<datatype>.description; use param.description](https://github.com/opctl/opctl/issues/898)
- [Docker images will only be pulled if using the `latest` tag (or untagged) or have not been pulled previously](https://github.com/opctl/opctl/issues/920)
- Go SDK models now use DataRef rather than PkgRef

### Fixed

- [vscode intellisense error](https://github.com/opctl/opctl/issues/615)

### Removed

- pkgs API endpoint; use data API endpoint
- Windows build; use linux build via WSL 2 instead

## [0.1.47] - 2021-01-22

### Added

- [Improve CLI prompts for username and password](https://github.com/opctl/opctl/issues/745)

### Fixed

- Dir initializer doesn't initialize more than one child entry

## [0.1.46] - 2021-01-04

### Fixed

- [container.envVars string double interpreted](https://github.com/opctl/opctl/issues/849)


cli/cli.go
package main

import (
	"context"
	"fmt"
	"os"
	"path/filepath"

	"github.com/appdataspec/sdk-golang/appdatapath"
	mow "github.com/jawher/mow.cli"
	"github.com/opctl/opctl/cli/internal/clioutput"
	"github.com/opctl/opctl/cli/internal/cliparamsatisfier"
	"github.com/opctl/opctl/cli/internal/dataresolver"
	"github.com/opctl/opctl/cli/internal/nodeprovider/local"
	"github.com/opctl/opctl/sdks/go/model"
	"github.com/opctl/opctl/sdks/go/opspec"
	"golang.org/x/term"
)

type cli interface {
	Run(args []string) error
}

func newCli(
	cliOutput clioutput.CliOutput,
) cli {

	cli := mow.App(
		"opctl",
		"Opctl is a free and open source distributed operation control system.",
	)
	cli.Version("v version", version)

	exitWith := func(successMessage string, err error) {
		if err == nil {
			if successMessage != "" {
				cliOutput.Success(successMessage)
			}
			mow.Exit(0)
		} else {
			cliOutput.Error(err.Error())

			if re, ok := err.(RunError); ok {
				mow.Exit(re.ExitCode)
			} else {
				mow.Exit(1)
			}
		}
	}

	perUserAppDataPath, err := appdatapath.New().PerUser()
	if err != nil {
		exitWith("", err)
	}

	containerRuntime := cli.String(
		mow.StringOpt{
			Desc:   "Runtime for opctl containers. Can be 'docker', 'k8s', or 'qemu' (experimental)",
			EnvVar: "OPCTL_CONTAINER_RUNTIME",
			Name:   "container-runtime",
			Value:  "docker",
		},
	)

	dataDir := cli.String(
		mow.StringOpt{
			Desc:   "Path of dir used to store opctl data",
			EnvVar: "OPCTL_DATA_DIR",
			Name:   "data-dir",
			Value:  filepath.Join(perUserAppDataPath, "opctl"),
		},
	)

	apiListenAddress := cli.String(
		mow.StringOpt{
			Desc:   "IP:PORT on which the API server will listen",
			EnvVar: "OPCTL_API_LISTEN_ADDRESS",
			Name:   "api-listen-address",
			Value:  "127.0.42.224:80",
		},
	)

	dnsListenAddress := cli.String(
		mow.StringOpt{
			Desc:   "IP:PORT on which the DNS server will listen",
			EnvVar: "OPCTL_DNS_LISTEN_ADDRESS",
			Name:   "dns-listen-address",
			Value:  "127.0.42.224:53",
		},
	)

	cliParamSatisfier := cliparamsatisfier.New(cliOutput)

	noColor := cli.BoolOpt("nc no-color", false, "Disable output coloring")

	cli.Before = func() {
		if *noColor {
			cliOutput.DisableColor()
		}
	}

	ctx, cancel := context.WithCancel(context.Background())

	cli.After = func() {
		cancel()
	}

	cli.Command("auth", "Manage auth for OCI image registries", func(authCmd *mow.Cmd) {
		authCmd.Command("add", "Add auth for an OCI image registry", func(addCmd *mow.Cmd) {
			addCmd.Spec = "RESOURCES [ -u=<username> ] [ -p=<password> ]"

			resources := addCmd.StringArg("RESOURCES", "", "Resources this auth applies to in the form of a host or host/path (e.g. docker.io)")
			username := addCmd.StringOpt("u username", "", "Username")
			password := addCmd.StringOpt("p password", "", "Password")

			addCmd.Action = func() {
				exitWith(
					"",
					auth(
						ctx,
						local.NodeConfig{
							APIListenAddress: *apiListenAddress,
							ContainerRuntime: *containerRuntime,
							DataDir:          *dataDir,
							DNSListenAddress: *dnsListenAddress,
						},
						model.AddAuthReq{
							Resources: *resources,
							Creds: model.Creds{
								Username: *username,
								Password: *password,
							},
						},
					),
				)
			}
		})
	})

	cli.Command("events", "Stream events", func(eventsCmd *mow.Cmd) {
		eventsCmd.Action = func() {
			exitWith(
				"",
				events(
					ctx,
					cliOutput,
					local.NodeConfig{
						APIListenAddress: *apiListenAddress,
						ContainerRuntime: *containerRuntime,
						DataDir:          *dataDir,
						DNSListenAddress: *dnsListenAddress,
					},
				),
			)
		}
	})

	cli.Command("ls", "List operations", func(lsCmd *mow.Cmd) {
		const dirRefArgName = "DIR_REF"
		lsCmd.Spec = fmt.Sprintf("[%v]", dirRefArgName)
		dirRef := lsCmd.StringArg(dirRefArgName, opspec.DotOpspecDirName, "Reference to dir ops will be listed from")

		lsCmd.Action = func() {
			exitWith(
				"",
				ls(
					ctx,
					cliParamSatisfier,
					local.NodeConfig{
						APIListenAddress: *apiListenAddress,
						ContainerRuntime: *containerRuntime,
						DataDir:          *dataDir,
						DNSListenAddress: *dnsListenAddress,
					},
					*dirRef,
				),
			)
		}
	})

	cli.Command("node", "Manage nodes", func(nodeCmd *mow.Cmd) {
		nodeCmd.Command("create", "Creates a node", func(createCmd *mow.Cmd) {
			createCmd.Action = func() {
				exitWith(
					"",
					nodeCreate(
						ctx,
						local.NodeConfig{
							APIListenAddress: *apiListenAddress,
							ContainerRuntime: *containerRuntime,
							DataDir:          *dataDir,
							DNSListenAddress: *dnsListenAddress,
						},
					),
				)
			}
		})

		nodeCmd.Command("delete", "Deletes a node. This is destructive! all node data including auth, caches, and operation state will be permanently removed.", func(deleteCmd *mow.Cmd) {
			deleteCmd.Action = func() {
				exitWith(
					"",
					nodeDelete(
						ctx,
						local.NodeConfig{
							APIListenAddress: *apiListenAddress,
							ContainerRuntime: *containerRuntime,
							DataDir:          *dataDir,
							DNSListenAddress: *dnsListenAddress,
						},
					),
				)
			}
		})

		nodeCmd.Command("kill", "Kills a node and any running operations. This is non destructive. All node data including auth, caches, and operation state will be retained.", func(killCmd *mow.Cmd) {
			killCmd.Action = func() {
				exitWith(
					"",
					nodeKill(
						ctx,
						local.NodeConfig{
							APIListenAddress: *apiListenAddress,
							ContainerRuntime: *containerRuntime,
							DataDir:          *dataDir,
							DNSListenAddress: *dnsListenAddress,
						},
					),
				)
			}
		})
	})

	cli.Command("op", "Manage ops", func(opCmd *mow.Cmd) {
		np, err := local.New(
			local.NodeConfig{
				APIListenAddress: *apiListenAddress,
				ContainerRuntime: *containerRuntime,
				DataDir:          *dataDir,
				DNSListenAddress: *dnsListenAddress,
			},
		)
		if err != nil {
			exitWith("", err)
		}

		node, err := np.CreateNodeIfNotExists(ctx)
		if err != nil {
			exitWith("", err)
		}

		dataResolver := dataresolver.New(
			cliParamSatisfier,
			node,
		)

		opCmd.Command("create", "Create an op", func(createCmd *mow.Cmd) {
			path := createCmd.StringOpt("path", opspec.DotOpspecDirName, "Path the op will be created at")
			description := createCmd.StringOpt("d description", "", "Op description")
			name := createCmd.StringArg("NAME", "", "Op name")

			createCmd.Action = func() {
				exitWith(
					"",
					opspec.Create(
						filepath.Join(*path, *name),
						*name,
						*description,
					),
				)
			}
		})

		opCmd.Command("install", "Install an op", func(installCmd *mow.Cmd) {
			path := installCmd.StringOpt("path", opspec.DotOpspecDirName, "Path the op will be installed at")
			opRef := installCmd.StringArg("OP_REF", "", "Op reference (either `relative/path`, `/absolute/path`, `host/path/repo#tag`, or `host/path/repo#tag/path`)")
			username := installCmd.StringOpt("u username", "", "Username used to auth w/ the pkg source")
			password := installCmd.StringOpt("p password", "", "Password used to auth w/ the pkg source")

			installCmd.Action = func() {
				exitWith(
					"",
					opInstall(
						ctx,
						dataResolver,
						*opRef,
						*path,
						&model.Creds{
							Username: *username,
							Password: *password,
						},
					),
				)
			}
		})

		opCmd.Command("kill", "Kill an op", func(killCmd *mow.Cmd) {
			opID := killCmd.StringArg("OP_ID", "", "Id of the op to kill")

			killCmd.Action = func() {
				exitWith(
					"",
					node.KillOp(
						ctx,
						model.KillOpReq{
							OpID:       *opID,
							RootCallID: *opID,
						},
					),
				)
			}
		})

		opCmd.Command("validate", "Validate an op", func(validateCmd *mow.Cmd) {
			opRef := validateCmd.StringArg("OP_REF", "", "Op reference (either `relative/path`, `/absolute/path`, `host/path/repo#tag`, or `host/path/repo#tag/path`)")

			validateCmd.Action = func() {
				exitWith(
					fmt.Sprintf("%v is valid", *opRef),
					opValidate(
						ctx,
						dataResolver,
						*opRef,
					),
				)
			}
		})
	})

	cli.Command("run", "Start and wait on an op", func(runCmd *mow.Cmd) {
		args := runCmd.StringsOpt("a", []string{}, "Explicitly pass args to op in format `-a NAME1=VALUE1 -a NAME2=VALUE2`")
		argFile := runCmd.StringOpt("arg-file", filepath.Join(opspec.DotOpspecDirName, "args.yml"), "Read in a file of args in yml format")
		noProgress := runCmd.BoolOpt("no-progress", !term.IsTerminal(int(os.Stdout.Fd())), "Disable live call graph for the op")
		opRef := runCmd.StringArg("OP_REF", "", "Op reference (either `relative/path`, `/absolute/path`, `host/path/repo#tag`, or `host/path/repo#tag/path`)")

		runCmd.Action = func() {
			exitWith(
				"",
				run(
					ctx,
					cliOutput,
					cliParamSatisfier,
					local.NodeConfig{
						APIListenAddress: *apiListenAddress,
						ContainerRuntime: *containerRuntime,
						DataDir:          *dataDir,
						DNSListenAddress: *dnsListenAddress,
					},
					*args,
					*argFile,
					*opRef,
					*noProgress,
				),
			)
		}
	})

	cli.Command("self-update", "Update opctl", func(selfUpdateCmd *mow.Cmd) {
		selfUpdateCmd.Action = func() {
			exitWith(
				selfUpdate(
					ctx,
					local.NodeConfig{
						APIListenAddress: *apiListenAddress,
						ContainerRuntime: *containerRuntime,
						DataDir:          *dataDir,
						DNSListenAddress: *dnsListenAddress,
					},
				),
			)
		}
	})

	cli.Command("ui", "Open the opctl web UI and mount a reference.", func(uiCmd *mow.Cmd) {
		const mountRefArgName = "MOUNT_REF"
		uiCmd.Spec = fmt.Sprintf("[%v]", mountRefArgName)
		mountRefArg := uiCmd.StringArg(mountRefArgName, ".", "Reference to mount (either `relative/path`, `/absolute/path`, `host/path/repo#tag`, or `host/path/repo#tag/path`)")

		uiCmd.Action = func() {
			exitWith(
				"Opctl web UI opened!",
				ui(
					ctx,
					cliParamSatisfier,
					local.NodeConfig{
						APIListenAddress: *apiListenAddress,
						ContainerRuntime: *containerRuntime,
						DataDir:          *dataDir,
						DNSListenAddress: *dnsListenAddress,
					},
					*apiListenAddress,
					*mountRefArg,
				),
			)
		}
	})

	return cli
}
